### ISSUE TITLE
[RFE] If both an iGPU and discrete GPU are detected, offer the option to rely solely on the iGPU

### ISSUE THREAD

**Issue by kirbyfan64** [rhinstaller/anaconda#1824]
e.g. by blacklisting the external GPU.

This seems like a rather odd request, but it comes from a long history of trying to help out others with NVIDIA cards on the Fedora Discord. Bugs related to nouveau aren't uncommon, whether the installer fails to start, or it successfully installs but won't boot.

A lot of the times, if the have no plans on using the NVIDIA card from Fedora (e.g. they're dual-booting and only use it on Windows), I'll walk them through blacklisting nouveau, and everything "just works" with puppies and kittens and flowers.

Now, here's the idea:

- The installer boot menu would have an additional option, either "Boot without a discrete GPU" / "Boot using only an integrated GPU", which will boot normally but blacklist all the normal discrete GPU kernel modules, or even just a "Boot without nouveau" (the former would probably seem less biased).
- In addition, during installation, there would be an option that would set the GRUB config to blacklist whichever kernel modules, e.g. "Rely solely on the system's integrated GPU". If the user booted with the "only integrated GPU" (or whatever name) boot entry, then this check box / option would be highlighted and maybe selected by default.

The amount of troubles that this would avoid for many people would be absolutely amazing.

**Comment by fepitre** [comment]
IMO it's a good idea. We can go further by also adding a mechanism for letting the user set specific grub options. I have already done things like that as a WIP few weeks ago so if the upstream is ok I can manage to implement it.

**Comment by jkonecny12** [comment]
If I understand it correctly this is not something what Anaconda should or can do. If you are talking about booting up installation with a mod blacklist than this should be done by [lorax](https://github.com/weldr/lorax) templates or maybe fedora grub configurations and ideally with knowledge of the bootloader developers. From what I understand there won't be any change on the Anaconda side.

@martinezjavier @jhlavac what do you think? What would be best place for doing this change?

**Comment by M4rtinK** [comment]
@kirbyfan64 Indeed, the first point would most likely have to be addressed in the static initial boot menu generated by Lorax. 

The second option would have to be done by Anaconda at runtime, but I wonder about a few things:
- how to reliably detect we are running on a hybrid graphics system ?
- once we detect we are on such a system, how to find out which module names to blacklist ?
- this would have to be an option, but where to put it ?
 - the main storage spoke has some bootloader related settings, but I'm not sure graphics related options really fit there

@fepitre Cool - yeah, I think that could be useful. But yet again - I wonder where is the best place in the UI for such a setting.

Maybe this even warrants a new "bootloader" spoke, where these options could be placed. We have recently changed the Hub layout, so an additional screen should fit.

**Comment by kirbyfan64** [comment]
> * how to reliably detect we are running on a hybrid graphics system ?

The devices and driver names should all be in /proc/bus/pci, which can be seen in `lspci -vv`. If there are >=2 devices with the video PCI class, then there's hybrid graphics.

> * once we detect we are on such a system, how to find out which module names to blacklist ?

See above, or just hardcode, since really the only discrete graphics drivers right now are amdgpu and nouveau.

* this would have to be an option, but where to put it ?

I guess where @fepitre mentioned?

**Comment by M4rtinK** [comment]
After some more digging in the Anaconda source code, it turns out we already support writing a kernel module blacklist:

```python
def _writeModuleBlacklist(self):
    """Copy modules from modprobe.blacklist=<module> on cmdline to
    /etc/modprobe.d/anaconda-blacklist.conf so that modules will
    continue to be blacklisted when the system boots.
    """
    if "modprobe.blacklist" not in flags.cmdline:
        return

    util.mkdirChain(util.getSysroot() + "/etc/modprobe.d")
    with open(util.getSysroot() + "/etc/modprobe.d/anaconda-blacklist.conf", "w") as f:
        f.write("# Module blacklists written by anaconda\n")
        for module in flags.cmdline["modprobe.blacklist"].split():
            f.write("blacklist %s\n" % module)
```

(from https://github.com/rhinstaller/anaconda/blob/master/pyanaconda/payload/__init__.py#L765)

So if the early bootloader menu "IGP only" option sets the `modprobe.blacklist` boot option to blacklist the `amdgpu` and `noveau` (I *think* this boo option is actually only handled by Anaconda and the modules need to be also explicitly disabled by the proper kernel mechanisms for the kernel being booted) then Anaconda will pick it up from the boot options and write it to a config file in `/etc/modprobe.d`. That should make sure the module keeps to be blacklisted also on the installed system.

**Comment by fepitre** [comment]
> @fepitre Cool - yeah, I think that could be useful. But yet again - I wonder where is the best place in the UI for such a setting.
> 
> Maybe this even warrants a new "bootloader" spoke, where these options could be placed. We have recently changed the Hub layout, so an additional screen should fit.

@M4rtinK : Yes indeed, it would be great to have a spoke for that and editing grub options to replace, at least, the defaults in pyanaconda/bootloader/grub2.py:write_defaults

I will do a POC and show you soon.

**Comment by M4rtinK** [comment]
> @M4rtinK : Yes indeed, it would be great to have a spoke for that and editing grub options to replace, at least, the defaults in pyanaconda/bootloader/grub2.py:write_defaults
> 
> I will do a POC and show you soon.
Cool! :)

**Comment by poncovka** [comment]
Thank you for reporting this issue! The issue tracker will be closed on 3 June 2019. Please make sure that your issue is also reported at https://bugzilla.redhat.com/. We are handling a high number of issues and many of them involve other components. Bugzilla allows us to better process your issues and collaborate on them with other teams. We are sorry for the inconvenience.
